### ISSUE TITLE
Failed to reproduce the results

### ISSUE THREAD

**Issue by EdwardKSG** [PurduePAML/K-ARM_Backdoor_Optimization#1]
This is an interesting work, by reading the paper. Therefore, I tried to run and test the code. Since README says "The default settings of the parameters in this repo can achieve 90% detection accuracy on TrojAI round3 training and testing datasets", I ran the code on TrojAI round 3 test set but only achieved around 71% accuracy, mainly because of vast false negatives (i.e. many trojan models were not detected). 

Every stage contributed to the false negatives: 
1.  A significant number of trojan models were filtered out by pre-screening.
2. It didn't manage to optimize a small-enough trigger on many trojan models. The final trigger sizes even stayed at the inial value 1e+10, which means they were not optimized at all.
3. Similar to 2, during the symmetry check, the optimizer also gave a trigger size of 1e+10, and thus the ratio between the trigger sizes of both direction is 1, which will not pass the symmetry check.

From the above observations, it seems like there are problems with the pre-screening phase and the trigger optimization. Do you think they were caused by some sub-optimal hyper-parameters?

**Comment by SolidShen** [comment]
> 

Hi, thanks for your interest of our work.  Our optimization component still follows the Neural Cleanse optimization which mainly handles patch(polygon) triggers. However, TrojAI R3 dataset has two types of triggers: polygon and filter(Instagram) triggers. 90% accuracy here means the detection accuracy on backdoor models with polygon triggers and benign models.  For your testing results, you can first check the 'model-id' of your failure cases and check their corresponding 'trigger_type' in the METADATA.csv file to see if most of them fall into the type of Instagram trigger.

If it is the case, you can refer line 2379-2387 in the following repo: 'https://github.com/naiyeleo/ABS/blob/master/TrojAI_competition/round3/abs_pytorch1_color_simple_filter4_10_5_determinism.py', which is a simple method to handle filter triggers. 

If not, we can further discuss if it's caused by some hyper-parameters settings. 

Thanks,
Guangyu

**Comment by EdwardKSG** [comment]
I used the METADATA file and simply ignored the ones with Instagram filters, then evaluated on the remaining models. For round 3 test set, the accuracy reached 91.7% which is quite impressive, with a true positive rate of 79.2% and true negative rate of 97.9%.  Do these stats look reasonable? The TN rate is very high, also since I removed the models with Instagram filters, the remaining dataset is not balanced anymore (roughly 2/3 of the models are benign). These lead to a good accuracy albeit a relatively humble TP rate.

I tried the same (i.e. removing models with Instagram filters) on round 4, but the TP rate is around 50%, which is a little bit low? The TN rate is still very good. Adjusting the trigger size bound and the symmetry ratio bound can help slightly improve TP rate to above 60%, but meanwhile significantly worsens the TN rate. Also, for many trojan models with patch triggers, the optimizer didn’t find a good trigger (the mask size stayed at 1e+10), does it mean I should increase the number of max round on an arm, or probably the total number of steps?

Also another question (sorry for bothering you so much): Since your code produces a definite result (either benign or trojan, not softmax logits), how was the cross entropy loss and AUC computed? Did you give a fixed probability (e.g. 0.9 when the results shows trojan, 0.1 when it is benign)?

**Comment by SolidShen** [comment]
> I used the METADATA file and simply ignored the ones with Instagram filters, then evaluated on the remaining models. For round 3 test set, the accuracy reached 91.7% which is quite impressive, with a true positive rate of 79.2% and true negative rate of 97.9%. Do these stats look reasonable? The TN rate is very high, also since I removed the models with Instagram filters, the remaining dataset is not balanced anymore (roughly 2/3 of the models are benign). These lead to a good accuracy albeit a relatively humble TP rate.
> 
> I tried the same (i.e. removing models with Instagram filters) on round 4, but the TP rate is around 50%, which is a little bit low? The TN rate is still very good. Adjusting the trigger size bound and the symmetry ratio bound can help slightly improve TP rate to above 60%, but meanwhile significantly worsens the TN rate. Also, for many trojan models with patch triggers, the optimizer didn’t find a good trigger (the mask size stayed at 1e+10), does it mean I should increase the number of max round on an arm, or probably the total number of steps?
> 
> Also another question (sorry for bothering you so much): Since your code produces a definite result (either benign or trojan, not softmax logits), how was the cross entropy loss and AUC computed? Did you give a fixed probability (e.g. 0.9 when the results shows trojan, 0.1 when it is benign)?

I think the round3 results are reasonable. For round4, there are several differences. For examples, there is no global trigger anymore, all triggers are label-specific. Besides, there is a new type of trigger called spatial trigger. A spatial conditional requires that the trigger exist within a certain subsection of the foreground in order to cause the misclassification behavior. You can refer more in [https://pages.nist.gov/trojai/docs/data.html#round-4](url). Therefore, I guess you might need to fine-tune some of the parameters a lit bit to get a better result.  For example, initial values for trigger mask and pattern, probability for epsilon greedy and max steps, etc.

Regarding to your last question, for round 1-4, we just simply give a fixed probability based on the inversion result. This value can be different for different type of triggers. We tune it on the training set.  For later NLP rounds(r8,r9), we train a classifier(decision tree or random forest) based on some features of the inverted trigger(size, asr, etc). You can certainly apply the similar strategy for the vision models. Based on our observation, training classifier usually gives you a slightly better detection accuracy compared on the naive rule-based method.

**Comment by SolidShen** [comment]
Closed since the issue is not active for months. Please reopen it if the issue remains in the code base.

**Comment by SolidShen** [comment]
Closed since the issue is not active for months. Please kindly reopen it if the issue remains in the code base.
